### PR TITLE
Create a `<TokenContext/>`  for storing token state with its setter and handleSubmit

### PR DESCRIPTION
**Pull request by cmarin001**
This pull request introduces the creation of a `<TokenContext/>` within our application. The purpose of this new context is to efficiently manage the state of various tokens along with their corresponding setter functions and to handle submission events. This implementation aims to streamline state management and enhance the overall functionality of our app's token-related features

### CODE REVIEW

File: src/routes/people.tsx [hunk 1]
```diff
@@ -6,63 +6,23 @@ import { Palette } from "@src/pages/people/outlets/palette";
...
-          element={<Surfaces token={token} handleSubmit={handleSubmit} />}
-        />
-        <Route
-          path="lines"
-          element={<Lines token={token} handleSubmit={handleSubmit} />}
-        />
-      </Route>
-      <Route path="/*" element={<ErrorPage />} />
-    </Routes>
+    <TokenProvider>
```
Comment (carin91) [inline src/routes/people.tsx:13]: Are you sure that any other component will not need this context?, i think so, however, confirm it please.

File: src/context/TokenContext/index.tsx [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  );
+
+  return (
+    <TokenContext.Provider value={{ token, handleSubmit }}>
+      {children}
+    </TokenContext.Provider>
+  );
+};
+
+export { TokenContext, TokenProvider };
```
Comment (carin91) [inline src/context/TokenContext/index.tsx:58]: I would create a types.ts file for at least defining the handleSubmit props interface.

File: src/pages/people/outlets/lines/interface.tsx [hunk 3]
```diff
@@ -31,7 +34,11 @@ type FormType = "spinner" | "link" | "default";
 interface IRenderForm {
   formType: string;
   selectedTab: string;
-  handleSubmit: (props: IHandleSubmitProps) => void;
+  handleSubmit: (
```
Comment (carin91) [inline src/pages/people/outlets/lines/interface.tsx:37]: Does the strokes interface need to import handleSubmit. I guess that you need it to pass as prop to its forms, however, the forms can import it directly from the context, doesn't it? This applies to every form (text, surface, stroke).

**Review by carin91 — CHANGES_REQUESTED** [review]

File: src/App.tsx [hunk 4]
```diff
@@ -10,7 +10,7 @@ import {
 import { ErrorPage } from "@components/layout/ErrorPage";
 import { Home } from "@pages/home";
 import { GlobalStyles } from "@styles/global";
-import AppContextProvider from "./context";
+import AppContextProvider from "./context/AppContext";
```
[the remark this answers is not in the document]
Comment (cmarin001) [reply src/App.tsx]: solved

File: src/pages/people/outlets/lines/form/RenderLinesContentForm/index.tsx [hunk 5]
```diff
@@ -14,7 +13,11 @@ import { IPeopleMessage } from "../../../types/people.types";
 
 interface RenderLinesContentFormProps {
   formType: string;
-  handleSubmit: (props: IHandleSubmitProps) => void;
+  handleSubmit: (
```
[the remark this answers is not in the document]
Comment (cmarin001) [reply src/pages/people/outlets/lines/form/RenderLinesContentForm/index.tsx:16]: solved

File: src/routes/people.tsx (hunk 1 above)
Comment (cmarin001) [reply src/routes/people.tsx:13]: Currently, only people need this to visualize the changes. If, in the future, others require it, adjustments will be made to accommodate that need

File: src/context/TokenContext/index.tsx (hunk 2 above)
Comment (cmarin001) [reply src/context/TokenContext/index.tsx:58]: solved

File: src/pages/people/outlets/lines/interface.tsx (hunk 3 above)
Comment (cmarin001) [reply src/pages/people/outlets/lines/interface.tsx:37]: solved

File: src/context/TokenContext/index.tsx [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  token: typeof inube;
+  handleSubmit: (
+    domain: string,
+    block: string,
+    tokenUpdate: typeof inube
+  ) => void;
+}
+
+const defaultTokenValue: ITokenState = {
+  token: presente,
```
[the remark this answers is not in the document]
Comment (cmarin001) [reply src/context/TokenContext/index.tsx:20]: solved

File: src/context/TokenContext/index.tsx [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+const TokenContext = createContext<ITokenState>(defaultTokenValue);
+
+interface ITokenProviderProps {
+  children: React.ReactNode;
+}
+
+const TokenProvider = (props: ITokenProviderProps) => {
+  const { children } = props;
+  const [token, setToken] = useState(presente);
```
[the remark this answers is not in the document]
Comment (cmarin001) [reply src/context/TokenContext/index.tsx:32]: solved

File: src/context/TokenContext/index.tsx [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        ...token.color,
+        [block]: { ...tokenUpdate },
+      };
+
+      const updatedToken = {
+        ...token,
+        [domain]: { ...updatedTokenColor },
+      };
+
+      setToken(updatedToken);
```
[the remark this answers is not in the document]
Comment (cmarin001) [reply src/context/TokenContext/index.tsx:46]: solved

File: src/pages/people/outlets/lines/interface.tsx (hunk 3 above)
Comment (carin91) [reply src/pages/people/outlets/lines/interface.tsx]: Pending:
As you can see in the example below you could useContext in the children which need to read it.

https://www.w3schools.com/react/react_usecontext.asp
image
In this case the handleSubmit and the token are required only on the forms. You read it in there to create the theme context and also you PUT the token with the handleSubmit, you are assigning an unnecessary responsibility to the  components for the tabs. I propose changes as presented below.

image
image
image

And... import the handleSubmit, token, and IhandleSubmitProps into the RenderSurfaceContentForm and the other forms.

Comment (carin91) [reply src/pages/people/outlets/lines/interface.tsx]: Pending:
As you can see in the example below you could useContext in the children which need to read it.

https://www.w3schools.com/react/react_usecontext.asp
image
In this case the handleSubmit and the token are required only on the forms. You read it in there to create the theme context and also you PUT the token with the handleSubmit, you are assigning an unnecessary responsibility to the  components for the tabs. I propose changes as presented below.

image
image
image

And... import the handleSubmit, token, and IhandleSubmitProps into the RenderSurfaceContentForm and the other forms.

File: src/pages/people/outlets/lines/index.tsx [hunk 9]
```diff
@@ -1,12 +1,11 @@
...
-
-import type { IPeopleColorProps } from "src/routes/people";
 import { linesFormsConfig } from "./config/lines.config";
 import { linesTabsConfig } from "./config/linesTabs.config";
+import { TokenContext } from "@src/context/TokenContext";
 
-function Lines(props: IPeopleColorProps) {
-  const { token, handleSubmit } = props;
+function Lines() {
+  const { token, handleSubmit } = useContext(TokenContext);
```
Comment (carin91) [inline src/pages/people/outlets/lines/index.tsx:8]: The lines index and interface do not use this information, the useContext should be used someplace else.

File: src/pages/people/outlets/lines/interface.tsx [hunk 10]
```diff
@@ -10,13 +10,13 @@ import { PageTitle } from "@components/PageTitle";
...
 import { StyledContainer, StyledTabsContainer } from "./styles";
 import { peopleOptionsConfig } from "../options/config/people.config";
-import { IHandleSubmitProps } from "@src/routes/people";
 import { linesTabsConfig } from "./config/linesTabs.config";
 
 import { linesFormsConfig } from "./config/lines.config";
 import { RenderLinesContentForm } from "./form/RenderLinesContentForm";
 import { RenderLinesWithSpinnerForm } from "./form/RenderLinesWithSpinnerForm";
 import { RenderLinesWithLinkForm } from "./form/RenderLinesWithLinkForm";
+import { IHandleSubmitProps } from "@src/context/TokenContext/types";
```
Comment (carin91) [inline src/pages/people/outlets/lines/interface.tsx:19]: The Strokes / interface does not use this information, the IHandleSubmitProps should be used someplace else (in the forms).

File: src/pages/people/outlets/palette/Palette.stories.tsx [hunk 11]
```diff
@@ -1,8 +1,8 @@
 import { StoryFn } from "@storybook/react";
 import { BrowserRouter } from "react-router-dom";
 import { Palette } from "./index";
+import { TokenContext } from "@src/context/TokenContext";
```
Comment (carin91) [inline src/pages/people/outlets/palette/Palette.stories.tsx:4]: i don't get it. Why the palette.stories needs the context? i supposed that the storybook was not design for that. It means that you need to pass the database information? Are you sure of it?

File: src/pages/people/outlets/palette/index.tsx [hunk 12]
```diff
@@ -1,12 +1,12 @@
-import { useEffect, useState } from "react";
+import { useContext, useEffect, useState } from "react";
```
Comment (carin91) [inline src/pages/people/outlets/palette/index.tsx:1]: palette / index does not need the tokens information for anything. Place the useContext in the forms

File: src/pages/people/outlets/surfaces/Surfaces.stories.tsx [hunk 13]
```diff
@@ -1,7 +1,7 @@
 import { StoryFn } from "@storybook/react";
 import { BrowserRouter } from "react-router-dom";
 import { Surfaces } from "./index";
-import { IPeopleColorProps } from "@src/routes/people";
+import { TokenContext } from "@src/context/TokenContext";
```
Comment (carin91) [inline src/pages/people/outlets/surfaces/Surfaces.stories.tsx:4]: These lines need to be reevaluated based on the revision of the storybook use.

File: src/pages/people/outlets/surfaces/form/RenderContentFormSurface/index.tsx [hunk 14]
```diff
@@ -3,13 +3,13 @@ import { RenderSurfaceContentFormUI } from "./interface";
...
 import { inube } from "@inube/design-system";
 
-import { IHandleSubmitProps } from "@src/routes/people";
 import { Appearance } from "@src/components/feedback/SendingInformation/types";
 import { getTokenColor } from "@src/components/cards/TokenColorCard/styles";
 import {
   surfaceFormsConfig,
   surfaceMessagesConfig,
 } from "../../config/surface.config";
+import { IHandleSubmitProps } from "@src/context/TokenContext/types";
```
Comment (carin91) [inline src/pages/people/outlets/surfaces/form/RenderContentFormSurface/index.tsx:12]: The forms and its children are the components which need to know about the handleSubmit and the tokens, so, the useContext must be placed here and the children get this information through props or the themeProvider.

File: src/pages/people/outlets/surfaces/index.tsx [hunk 15]
```diff
@@ -1,11 +1,11 @@
-import { useState } from "react";
+import { useContext, useState } from "react";
```
Comment (carin91) [inline src/pages/people/outlets/surfaces/index.tsx:1]: The surface / index does not need the tokens information for anything. Place the useContext in the form

File: src/pages/people/outlets/surfaces/interface.tsx [hunk 16]
```diff
@@ -11,9 +11,9 @@ import { StyledContainer, StyledTabsContainer } from "./styles";
 import { peopleOptionsConfig } from "../options/config/people.config";
 import { surfaceTabsConfig } from "./config/surfaceTabs.config";
 import { RenderSurfaceContentForm } from "./form/RenderContentFormSurface";
-import { IHandleSubmitProps } from "@src/routes/people";
 import { surfaceFormsConfig } from "./config/surface.config";
 import { Appearance } from "@src/components/feedback/SendingInformation/types";
+import { IHandleSubmitProps } from "@src/context/TokenContext/types";
```
Comment (carin91) [inline src/pages/people/outlets/surfaces/interface.tsx:16]: The Surface/ interface does not use this information, the IHandleSubmitProps should be used someplace else (in the forms).

File: src/pages/people/outlets/texts/Texts.stories.tsx [hunk 17]
```diff
@@ -1,9 +1,8 @@
 import { StoryFn } from "@storybook/react";
 import { BrowserRouter } from "react-router-dom";
-
-import { IPeopleColorProps } from "@src/routes/people";
-import { presente } from "@inube/design-system";
 import { Texts } from ".";
+import { TokenContext } from "@src/context/TokenContext";
```
Comment (carin91) [inline src/pages/people/outlets/texts/Texts.stories.tsx:4]: These lines need to be reevaluated based on the revision of the storybook use.

File: src/pages/people/outlets/texts/form/RenderTextContentForm/index.tsx [hunk 18]
```diff
@@ -4,8 +4,8 @@ import { inube } from "@inube/design-system";
 import { textFormsConfig, textMessagesConfig } from "../../config/text.config";
 import { IUsersMessage } from "@src/pages/privileges/outlets/users/types/users.types";
 import { Appearance } from "@src/components/cards/FieldsetColorCard/types";
-import { IHandleSubmitProps } from "@src/routes/people";
 import { getTokenColor } from "@src/components/cards/TokenColorCard/styles";
+import { IHandleSubmitProps } from "@src/context/TokenContext/types";
```
Comment (carin91) [inline src/pages/people/outlets/texts/form/RenderTextContentForm/index.tsx:8]: The forms and its children are the components which need to know about the handleSubmit and the tokens, so, the useContext must be placed here and the children get this information through props or the themeProvider.

File: src/pages/people/outlets/texts/index.tsx [hunk 19]
```diff
@@ -1,11 +1,11 @@
-import { useState } from "react";
+import { useContext, useState } from "react";
```
Comment (carin91) [inline src/pages/people/outlets/texts/index.tsx:1]: The Text/ index does not need the tokens information for anything. Place the useContext in the form

File: src/pages/people/outlets/texts/interface.tsx [hunk 20]
```diff
@@ -11,8 +11,8 @@ import { StyledContainer, StyledTabsContainer } from "./styles";
 import { peopleOptionsConfig } from "../options/config/people.config";
 import { textsTabsConfig } from "./config/textsTabs.config";
 import { RenderTextContentForm } from "./form/RenderTextContentForm";
-import { IHandleSubmitProps } from "@src/routes/people";
 import { textFormsConfig } from "./config/text.config";
+import { IHandleSubmitProps } from "@src/context/TokenContext/types";
```
Comment (carin91) [inline src/pages/people/outlets/texts/interface.tsx:15]: The Text/ interface does not use this information, the IHandleSubmitProps should be used someplace else (in the forms).

**Review by carin91 — CHANGES_REQUESTED** [review]

File: src/routes/people.tsx (hunk 1 above)
Comment (carin91) [reply src/routes/people.tsx]: As related in another comment, i think this is the only correct place for the tokenProvider. Why do you change it? Do you think that you will need the tokens outside people?

File: src/App.tsx [hunk 21]
```diff
@@ -52,8 +53,10 @@ function App() {
   }
   return (
     <AppContextProvider>
-      <GlobalStyles />
-      <RouterProvider router={router} />
+      <TokenProvider>
```
[the remark this answers is not in the document]
Comment (cmarin001) [reply src/App.tsx:56]: solved

File: src/pages/people/outlets/lines/interface.tsx (hunk 3 above)
Comment (cmarin001) [reply src/pages/people/outlets/lines/interface.tsx]: solved

File: src/pages/people/outlets/lines/index.tsx (hunk 9 above)
Comment (cmarin001) [reply src/pages/people/outlets/lines/index.tsx]: solved

File: src/pages/people/outlets/lines/interface.tsx (hunk 10 above)
Comment (cmarin001) [reply src/pages/people/outlets/lines/interface.tsx]: solved

File: src/pages/people/outlets/palette/index.tsx (hunk 12 above)
Comment (cmarin001) [reply src/pages/people/outlets/palette/index.tsx]: solved

File: src/pages/people/outlets/surfaces/form/RenderContentFormSurface/index.tsx (hunk 14 above)
Comment (cmarin001) [reply src/pages/people/outlets/surfaces/form/RenderContentFormSurface/index.tsx]: solved

File: src/pages/people/outlets/surfaces/index.tsx (hunk 15 above)
Comment (cmarin001) [reply src/pages/people/outlets/surfaces/index.tsx]: solved

File: src/pages/people/outlets/surfaces/interface.tsx (hunk 16 above)
Comment (cmarin001) [reply src/pages/people/outlets/surfaces/interface.tsx]: solved

File: src/pages/people/outlets/texts/form/RenderTextContentForm/index.tsx (hunk 18 above)
Comment (cmarin001) [reply src/pages/people/outlets/texts/form/RenderTextContentForm/index.tsx]: solved

File: src/pages/people/outlets/texts/interface.tsx (hunk 20 above)
Comment (cmarin001) [reply src/pages/people/outlets/texts/interface.tsx]: solved

**Review by carin91 — APPROVED** [review]

File: src/context/TokenContext/index.tsx [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+
+const TokenContext = createContext<ITokenContextProps>(defaultTokenValue);
+
+interface ITokenProviderProps {
+  children: React.ReactNode;
+}
+
+const tokenReducer = (state: typeof inube, action: TokenActions) => {
+  switch (action.type) {
+    case actionTypes.SET_TOKEN:
```
Comment (carin91) [inline src/context/TokenContext/index.tsx:24]: This action is not in use, i guess this is for future implementation, i would evaluate to delete it.

Comment (cmarin001) [reply src/context/TokenContext/index.tsx:24]: we keep it for future implementation

**Pull request merged**